proxy/envoy/config.go

```diff
@@ -524,25 +524,47 @@ func buildOutboundHTTPRoutes(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.No
 //
 // Temporary workaround is to add a listener for each service IP that requires
 // TCP routing
+//
+// Connections to the ports of non-load balanced services are directed to
+// the connection's original destination. This avoids costly queries of instance
+// IPs and ports, but requires that ports of non-load balanced service be unique.
 func buildOutboundTCPListeners(mesh *proxyconfig.ProxyMeshConfig, services []*model.Service) (Listeners, Clusters) {
 	tcpListeners := make(Listeners, 0)
 	tcpClusters := make(Clusters, 0)
+
+	var originalDstCluster *Cluster
 	for _, service := range services {
 		if service.External() {
 			continue // TODO TCP external services not currently supported
 		}
 		for _, servicePort := range service.Ports {
 			switch servicePort.Protocol {
 			case model.ProtocolTCP, model.ProtocolHTTPS, model.ProtocolMONGO:
-				cluster := buildOutboundCluster(service.Hostname, servicePort, nil)
-				route := buildTCPRoute(cluster, []string{service.Address})
-				config := &TCPRouteConfig{Routes: []*TCPRoute{route}}
-				listener := buildTCPListener(config, service.Address, servicePort.Port, servicePort.Protocol)
-				tcpClusters = append(tcpClusters, cluster)
-				tcpListeners = append(tcpListeners, listener)
+				if service.Address != "" {
+					cluster := buildOutboundCluster(service.Hostname, servicePort, nil)
+					route := buildTCPRoute(cluster, []string{service.Address})
+					config := &TCPRouteConfig{Routes: []*TCPRoute{route}}
+					listener := buildTCPListener(
+						config, service.Address, servicePort.Port, servicePort.Protocol)
+					tcpClusters = append(tcpClusters, cluster)
+					tcpListeners = append(tcpListeners, listener)
+				} else {
+					if originalDstCluster == nil {
```

Comment:
the order of Services is non-det and this seems to pick the first one that has no address. Can we change the logic so that we pick one original destination cluster for each servicePort.Port (by sorting services by name)?
We also need to be careful with merging these listeners with HTTP listeners as they can collide.